### PR TITLE
Networkx version of construct function

### LINKED ISSUE #2: Combine all the network functions into a class

### LINKED ISSUE #3: Use an environment variable to pick backend
Set `PP_NETLIB_BACKEND` to one of graphtool (default), networkx or cugraph. This will be selected on package import.
When `.load()` is called, automatically convert to desired type

### PR DESCRIPTION
**Pull request by bruhad-dave**
Closes #2, closes #3

### CODE REVIEW

File: tests/test_backend_selector.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    parser.add_argument("--set_with_python", action="store_true", help="Use flag to set backend with od.environ.")
+    args = parser.parse_args()
+    set_with_py = args.set_with_python
+    return set_with_py
+
+def __init__():
+    vertex_labels = ["s1", "s2", "s3", "s4", "s5"]
+    sample_edge_list = [(0, 1), (1, 2), (2, 3), (3, 4), (4, 0)]
+    weights = [0.1, 0.1, 0.5, 0.7, 0.2]
+    sample_weighted_edge_list = [(0, 1, 0.1), (1, 2, 0.1), (2, 3, 0.5), (3, 4, 0.7), (4, 0, 0.2)]
```
Comment (johnlees) [inline tests/test_backend_selector.py:23]: Did you fix the issue from before with missing some of the edges? It looks like the edge list here is a lot smaller than before

Comment (bruhad-dave) [reply tests/test_backend_selector.py:23]: yep, that issue is fixed, but it is in a different branch. should i make those changes here?

File: pp_netlib/functions.py [hunk 2]
```diff
@@ -1,238 +1,124 @@
 ########################
 ####   .CONSTRUCT   ####
 ########################
-def initial_graph_properties(ref_list, query_list):
-    """Initial processing of sequence names for
-    network construction.
+import scipy
+import pandas as pd
+import graph_tool.all as gt
+import networkx as nx
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:7]: moved conditional imports to inside of the __init__ function in the Network class; additionally moved import into their specific construct functions (possibly redundant but just for insurance)

File: pp_netlib/functions.py [hunk 3]
```diff
@@ -1,238 +1,124 @@
...
+                graph.add_edge(network_data["source"][i], network_data["destination"][i], weight=network_data["weights"][i])
+        else:
+            graph.add_edges_from(network_data.values)
+
+    ##########################
+    #### SPARSE MAT INPUT ####
+    ##########################
+    elif isinstance(network_data, scipy.sparse.coo_matrix):
+        for i in range(len(vertex_labels)):
+            graph.add_edge(network_data.row[i], network_data.col[i], weight=network_data.data[i])
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:111]: done, used add_weighted_edges_from

File: pp_netlib/functions.py [hunk 4]
```diff
@@ -1,238 +1,124 @@
...
+
+    ########################
+    ####    DF INPUT    ####
+    ########################
+    if isinstance(network_data, pd.DataFrame):
+        network_data.columns = ["source", "destination"]
+        if weights is not None:
+            network_data["weights"] = weights
+            for i in range(len(vertex_labels)):
+                graph.add_edge(network_data["source"][i], network_data["destination"][i], weight=network_data["weights"][i])
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:102]: done, used add_weighted_edges_from

File: pp_netlib/functions.py [hunk 5]
```diff
@@ -1,238 +1,124 @@
...
+            graph.add_edge(network_data.row[i], network_data.col[i], weight=network_data.data[i])
+
+    ########################
+    ####   LIST INPUT   ####
+    ########################
+    elif isinstance(network_data, list):
+        if weights is not None:
+            src, dest = zip(*network_data)
+            for i in range(len(vertex_labels)):
+                graph.add_edge(src[i], dest[i], weight=weights[i])
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:120]: done, used add_weighted_edges_from

File: tests/test_backend_selector.py (hunk 1 above)
Comment (johnlees) [reply tests/test_backend_selector.py:23]: Yes, make them here

File: pp_netlib/functions.py [hunk 6]
```diff
@@ -1,238 +1,124 @@
...
+    return graph
+
+def construct_with_networkx(network_data, vertex_labels, use_gpu, weights = None):
+    ## initialise nx graph and add nodes
+    graph = nx.Graph()
+    graph.add_nodes_from(range(len(vertex_labels)))
+
+    ## add node labels
+    for i in range(len(vertex_labels)):
+        graph.nodes[i]["id"] = vertex_labels[i]
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/functions.py:92]: done, created a formatted list of nodes then added nodes and ids in same step

File: pp_netlib/network.py [hunk 7]
```diff
@@ -3,6 +3,9 @@
 import pandas as pd
 import scipy
 import graph_tool.all as gt
+import networkx as nx
```
[the remark this answers is not in the document]
Comment (bruhad-dave) [reply pp_netlib/network.py:6]: moved imports to inside __init__ and made them conditional to backend value

File: tests/test_backend_selector.py (hunk 1 above)
Comment (bruhad-dave) [reply tests/test_backend_selector.py:23]: added those changes

File: pp_netlib/network.py [hunk 8]
```diff
@@ -2,8 +2,8 @@
 import numpy as np
 import pandas as pd
 import scipy
-import graph_tool.all as gt
-import networkx as nx
+# import graph_tool.all as gt
+# import networkx as nx
```
Comment (johnlees) [inline pp_netlib/network.py:6]: Delete these
```suggestion
```

**Review by johnlees — CHANGES_REQUESTED** [review]
Two small final things:
- One deletion below
- `__pycache__` should not be committed. Run `git rm __pycache__` and commit. Then add the `.gitignore` file for python packages ([here](https://github.com/github/gitignore/blob/main/Python.gitignore)) and commit that

**Review by johnlees — APPROVED** [review]

**Comment by johnlees** [comment]
Happy for you to merge when tests have passed

**Pull request merged**